fe/endpoint_correction.py

```diff
@@ -0,0 +1,199 @@
+    lhs_xs,
+    rhs_xs):
+    """
+    Compute the BAR re-weighted end-point correction of converting an intractable core
+    restraint into a tractable RMSD-based orientational restraint.
+
+    Parameters
+    ----------
+    k_translation: float
+        Force constant of the translational restraint
+
+    k_rotation: float
+        Force constant of the rotational restraint
+
+    core_idxs: int np.array (C, 2)
+        Atom mapping between the two cores
+
+    core_params: float np.array (C, 2)
+        Bonded parameters of the intractable restraint
+
+    lhs_xs: np.array [T, N, 3]
+        Samples from the intractable left hand state, with the restraints turned on.
+
+    rhs_xs: np.array [T, N, 3]
+        Samples from the non-interacting, fully unrestrained right hand state.
+
+    beta: 1/kT
+        inverse kT
+
+    Returns
+    -------
+    4-tuple
+        4-tuple array of lhs delta_us, rhs delta_us, sampled rotations, and sampled translations.
+        The sign of the delta_us is defined to be consistent with that of pymbar. i.e. lhs_du measures
+        difference of u_rhs - u_lhs, using samples from lhs_xs. Whereas rhs_du measures difference of
+        u_lhs - u_rhs, using samples from rhs_xs.
+
+    """
+
+    box = np.eye(3) * 100.0
+    core_restr = functools.partial(
+        bonded.harmonic_bond,
+        bond_idxs=core_idxs,
+        params=core_params,
+        box=box,
+        lamb=None
+    )
+
+    # center of mass translational restraints
+    restr_group_idxs_a = core_idxs[:, 0]
+    restr_group_idxs_b = core_idxs[:, 1]
+
+    # disjoint sets
+    assert len(set(restr_group_idxs_a.tolist()).intersection(set(restr_group_idxs_b.tolist()))) == 0
+
+    translation_restr = functools.partial(
+        bonded.centroid_restraint,
+        group_a_idxs=restr_group_idxs_a,
+        group_b_idxs=restr_group_idxs_b,
+        params=None,
+        kb=k_translation,
+        b0=0.0,
+        box=box,
+        lamb=None
+    )
+
+    rotation_restr = functools.partial(
+        bonded.rmsd_restraint,
+        params=None,
+        group_a_idxs=restr_group_idxs_a,
+        group_b_idxs=restr_group_idxs_b,
+        k=k_rotation,
+        box=box,
+        lamb=None
+    )
+
+    # (ytz): delta_U is simplified to this expression as the rest of the hamiltonian is unaffected
+    # by a rigid translation/rotation
+    def delta_u_fn(x_t):
+        return translation_restr(x_t) + rotation_restr(x_t) - core_restr(x_t)
+
+    delta_u_batch = jax.jit(jax.vmap(delta_u_fn))
+
+    lhs_du = delta_u_batch(lhs_xs)
+
+    sample_size = rhs_xs.shape[0]
+    rotation_samples = sample_multiple_rotations(k_rotation, beta, sample_size)
+    covariance = np.eye(3)/(2*beta*k_translation)
+    translation_samples = np.random.multivariate_normal((0,0,0), covariance, sample_size)
+
+    rhs_xs_aligned = []
+    rhs_du = []
+
+    for x, r, t in zip(rhs_xs, rotation_samples, translation_samples):
+        # this can be batched as well, at some point.
```

Comment:
I think this loop is now the bottleneck, will look into batching this via vmap tomorrow!